// fusion-core/dist-browser-cjs/server-app.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = _default;

var _compose = require("./compose.js");

var _timing = _interopRequireWildcard(require("./plugins/timing"));

var _baseApp = _interopRequireDefault(require("./base-app"));

var _serverRenderer = _interopRequireDefault(require("./plugins/server-renderer"));

var _tokens = require("./tokens");

var _ssr = _interopRequireDefault(require("./plugins/ssr"));

var _serverContext = _interopRequireDefault(require("./plugins/server-context.js"));

var _appSymbol = require("./utils/app-symbol.js");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

/** Copyright (c) 2018 Uber Technologies, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * 
 */

/* eslint-env node */
function _default() {
  const Koa = require('koa');

  return class ServerApp extends _baseApp.default {
    constructor(el, render) {
      super(el, render);
      this.endpoints = new Map();
      this._app = new Koa();
      this._app.proxy = true;
      this.middleware(_serverContext.default);
      this.middleware((ctx, next) => {
        ctx[_appSymbol.appSymbol] = this;
        return next();
      });
      this.register(_timing.TimingToken, _timing.default);
      this.middleware((ctx, next) => {
        for (const [endpointPath, handler] of this.endpoints) {
          if (ctx.path === endpointPath) {
            return handler(ctx, next);
          }
        }

        return next();
      });
      this.middleware({
        element: _tokens.ElementToken,
        ssrDecider: _tokens.SSRDeciderToken,
        ssrBodyTemplate: _tokens.SSRBodyTemplateToken.optional,
        ssrShellTemplate: _tokens.SSRShellTemplateToken.optional
      }, (0, _ssr.default)(this.endpoints));
    }

    resolve() {
      this.middleware({
        timing: _timing.TimingToken,
        render: _tokens.RenderToken,
        ssrDecider: _tokens.SSRDeciderToken
      }, (0, _serverRenderer.default)(this));
      return super.resolve();
    }

    callback() {
      this.resolve();

      this._app.use((0, _compose.compose)(this.plugins));

      return this._app.callback();
    }

  };
}
